[2.7][Filesystem] Changed dumpFile to allow dumping to streams...

markchalloner opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #10018 |
| License | MIT |
| Doc PR | symfony/symfony-docs#5320 |

...as rename can only work where $oldname and $newname have the same streams wrappers.

Added tempnam, stream wrapper aware version of tempnam(), to do the heavy lifting.

markchalloner committed:
[Filesystem] added tempnam() stream wrapper aware version of PHP's native tempnam() and fixed dumpFile to allow dumping to streams

stof commented on src/Symfony/Component/Filesystem/Filesystem.php:
```diff
@@ -501,4 +558,34 @@ private function toIterator($files)
+     *
+     * @return string The filename hierarchy.
+     */
+    private function getHierarchy($filename)
```
hierarchy ? this is the first time I see this word used for a path component

markchalloner replied:
Includes the authority as well as the path.

STD 66 and RFC 3986 call it "hierarchical part" but hierarchy seemed nicer.

http://en.wikipedia.org/wiki/URI_scheme

Cheers

stof commented on src/Symfony/Component/Filesystem/Filesystem.php:
```diff
@@ -453,6 +453,63 @@ public function isAbsolutePath($file)
+        $limit = 10;
+        $schemeSeparator = '://';
+        $scheme = $this->getScheme($dir);
+        $hierarchy = $this->getHierarchy($dir);
```
rather than implementing these in 2 separate methods, each of them redoing the same splitting of the path, what about doing the logic determining both in the same place (either here or in a method returning ``array($scheme, $hierarchy)`` and using ``list($scheme, $hierarchy) = ...`` here) ?

stof commented on src/Symfony/Component/Filesystem/Filesystem.php:
```diff
@@ -501,4 +558,34 @@ private function toIterator($files)
+     *
+     * @param string $filename The filename to be parsed.
+     *
+     * @return string|bool The filename scheme if present or false.
```
please use ``string|null`` rather than ``string|false``

stof commented on src/Symfony/Component/Filesystem/Filesystem.php:
```diff
@@ -453,6 +453,63 @@ public function isAbsolutePath($file)
+
+            // If the file already exists restart the loop
+            // Use fopen instead of file_exists as some streams don't support stat
+            if (false !== @fopen($tmpFile, 'r')) {
```
you need to close the stream after opening it, otherwise you will be leaking stream resources

stof commented on src/Symfony/Component/Filesystem/Filesystem.php:
```diff
@@ -453,6 +453,63 @@ public function isAbsolutePath($file)
+            }
+
+            // Create the file, if unsuccessful then stream is not writable so set to false
+            // Use file_put_contents instead of touch as it support stream wrappers
```
typo: ``supports``

javiereguiluz commented:
In the current `Filesystem` component, all custom functions use the same name as the original PHP function (`copy`, `mkdir`, `touch`, etc.) Why should we change the name of the PHP function in this case (`tempnam()` -> `tempNam()`)?

staabm commented on src/Symfony/Component/Filesystem/Filesystem.php:
```diff
@@ -453,6 +453,66 @@ public function isAbsolutePath($file)
+
+            // If the file already exists restart the loop
+            // Use fopen instead of file_exists as some streams do not support stat
+            $handle = @fopen($tmpFile, 'r');
```
why don't you use `file_exists` (or `is_readable`)?

markchalloner replied:
@staabm. See comment just above :)

staabm replied:
arghs sry ;)

aitboudad commented on src/Symfony/Component/Filesystem/Filesystem.php:
```diff
@@ -501,4 +561,20 @@ private function toIterator($files)
+    private function getSchemeAndHierarchy($filename)
+    {
+        $schemeSeparator = '://';
+        $components = explode($schemeSeparator, $filename, 2);
```
we don't need  $schemeSeparator here 
```php
$components = explode('://', $filename, 2);
```

aitboudad commented on src/Symfony/Component/Filesystem/Filesystem.php:
```diff
@@ -453,6 +453,66 @@ public function isAbsolutePath($file)
+        // If no scheme or scheme is "file" create temp file in local filesystem
+        if (null === $scheme || 'file' === $scheme) {
+            $tmpFile = tempnam($hierarchy, $prefix);
+        }
```
we can simplify the below [if statements] (https://github.com/markchalloner/symfony/blob/ticket_10018/src/Symfony/Component/Filesystem/Filesystem.php#L475-L484) and remove $schemeSeparator too.
```php
// If no scheme or scheme is "file" create temp file in local filesystem
if (null === $scheme || 'file' === $scheme) {
    $tmpFile = tempnam($hierarchy, $prefix);

    return null === $scheme ? $tmpFile:$scheme.'://'.$tmpFile;
}
```

markchalloner replied:
@aitboudad. Thanks for the feedback.

markchalloner committed:
Fixed TOCTOU bug and added MockStream for tests

markchalloner commented:
I think I fixed a TOCTOU security vulnerability but can someone check:

https://github.com/markchalloner/symfony/commit/26cbdfbd49b708367e459b7328859bef3207486a#diff-e88ad8ede40599bea0592b1113c4452c

I also added a MockStream as none of the built-in ones were suitable for testing (see comments in testTempnamWith*SchemeFails).

markchalloner commented:
Rebased onto 2.8 in  #14970
